src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts

```diff
@@ -87,44 +93,50 @@ export class CategoryFacetValues {
     }, this.facet);
   }
 
-  private createFacetValueAtPath(path: string[]) {
-    const remainingPath = [...path];
-    const ultimateFacetValue = this.createFacetValueWithPath([...remainingPath]);
-    let facetValue = ultimateFacetValue;
-
-    while (remainingPath.length > 1) {
-      remainingPath.pop()
-      const parentFacetValue = this.findValueWithPath(remainingPath) || this.createFacetValueWithPath([...remainingPath]);
-      parentFacetValue.children.push(facetValue);
-      facetValue = parentFacetValue;
+  private getOrCreateFacetValueWithPath(path: string[]) {
+    const existingFacetValue = this.findValueWithPath(path);
+    if (existingFacetValue) {
+      return existingFacetValue;
     }
 
-    this.facetValues.push(facetValue);
-    return ultimateFacetValue;
-  }
-
-  public get(path: string[]) {
-    const facetValue = this.findValueWithPath(path);
-
-    if (facetValue) {
-      return facetValue;
-    }
+    const newFacetValue = this.createFacetValueWithPath(path);
+    const siblingValues = path.length === 1
+      ? this.facetValues
+      : this.getOrCreateFacetValueWithPath(path.slice(0, -1)).children;
 
-    const newFacetValue = this.createFacetValueAtPath(path);
+    siblingValues.push(newFacetValue)
     return newFacetValue;
   }
 
   private collapseHierarchyAtPathLevel(facetValues: CategoryFacetValue[], path: string[], level = 1) {
-    facetValues.forEach(facetValue => {
+    const collapsedFacetValues = facetValues
+      .filter(facetValue => {
+        const targetPath = path.slice(0, level);
+        return Utils.arrayEqual(targetPath, facetValue.path);
+      });
+
+    collapsedFacetValues.forEach(facetValue => {
       facetValue.state = FacetValueState.idle;
-      const targetPath = path.slice(0, level + 1)
-      facetValue.children = path[level] ? facetValue.children.filter(child => Utils.arrayEqual(targetPath, child.path)) : [];
-      this.collapseHierarchyAtPathLevel(facetValue.children, path, level + 1);
+      facetValue.children = this.collapseHierarchyAtPathLevel(facetValue.children, path, level + 1);
     });
+
+    return collapsedFacetValues;
+  }
+
+  private collapseHierarchyWithPath(path: string[]) {
+    this.facetValues = this.collapseHierarchyAtPathLevel(this.facetValues, [...path]);
```

Comment:
Is cloning the path still relevant? It doesn't seem like `collapseHierarchyAtPathLevel` modifies `path`.

Reply:
You're right!